vtex-apps/glovo-integration
# Feat/create records automatically

#### What problem is this solving?
Before this PR, when a client installed the app, they were required to create some records manually by sending POST request to certain endpoints for store menu updates and products records.
To give the the app a better UX, now this records are created automatically when needed, thus simplifying the initial process.

#### How to test it?
The app is linked is linked in this [workspace](https://glovo--ametllerorigenqa.myvtex.com/).
To test, you can make some changes in the stock for the store [here](https://glovo--ametllerorigenqa.myvtex.com/admin/inventory/?keyword=Aranja%20rosa%20-%20Aranja%20rosa&skuId=100).
You can try changing the stock for 'Aranja Rosa' in the 'Numancia' store (sales channel = 12).

To test the creation of a new store you can add a new store in the app settings using sales channel 12.
You can see the logs [here](https://splunk72.vtex.com/en-US/app/search/search?q=search%20index%3Dio_vtex_logs%20account%3Dametllerorigenqa%20workspace%3Dglovo%20app%3D%22vtex.glovo-integration%401.x%22&sid=1636127739.153978_0FC90BB0-9222-411B-B48E-3D77BD82C6AF&display.page.search.mode=smart&dispatch.sample_ratio=1&workload_pool=&earliest=-5m%40m&latest=now).

#### Changes
There seems to be a lot of changes (22 files! 😟). But don't panic!
Most of these changes are just because of changing `affiliate` to `store` to be more consistent with naming conventions in the app.

The important changes are the following:
1. Handling of Catalog Update event:
The new logic for this process can be found [here](https://github.com/vtex-apps/glovo-integration/blob/b04e80731de1490087af144e479988950399dc8c/node/utils.ts#L277-L441). Now we check if the product and store menu updates records exist and create them if they don't.
 
2. Creating/updating the Glovo menu:
When the client creates or updates the Glovo menu, we check if every store on the app settings has a `store menu updates` and create one for the ones that don't have it. You can see the logic for this [here](https://github.com/vtex-apps/glovo-integration/blob/b04e80731de1490087af144e479988950399dc8c/node/middlewares/saveGlovoMenu.ts#L20-L48)

3. Store menu updates record data structure:
The `store menu updates` record was an array with two elements, [0] for the previous update and [1] for the current one. This was changed to an object with two keys, `current` and `previous` as can be seen [here](https://github.com/vtex-apps/glovo-integration/blob/b04e80731de1490087af144e479988950399dc8c/node/typings.d.ts#L438-L441). This makes it clearer as to what update we are targeting in the code.

4. Simplify and remove unnecessary endpoints:
Because the user app did not created the records it needed automatically, it was required to have endpoints to create them. By making this process automatic, we are able to remove this endpoints as they are unnecessary.


Ps: we still have the documentation update (readme and changelog) pending.

#### Screenshots or example usage:
![image](https://user-images.githubusercontent.com/17585823/140540322-9c4d6a21-d288-4464-9fc2-c3823f07edfe.png)
